### PR TITLE
Integrating 4MC compression with Apache Beam

### PR DESCRIPTION
**Pull request by sundarrajanraman**
Integrating 4MC compression that supports splittable compressed file format (4mc) which lets us leverage the power of LZ4 and ZSTD algorithms.

It differs from other compressed file formats in a way that the output files can be read using HIVE partitioning. This PR allows Dataflow to write files in 4MC compression format.

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by johnjcasey** [comment]
run java precommit

File: sdks/java/core/build.gradle [hunk 1]
```diff
@@ -82,6 +83,7 @@ dependencies {
   implementation library.java.antlr_runtime
   implementation library.java.commons_compress
   implementation library.java.commons_lang3
+  implementation group: 'com.fing.fourmc', name: 'hadoop-4mc', version: '3.0.0'
```
Comment (johnjcasey) [inline sdks/java/core/build.gradle:86]: Please add this as a library, similar to the other libraries we import

File: sdks/java/core/build.gradle [hunk 2]
```diff
@@ -17,6 +17,7 @@
  */
 
 plugins { id 'org.apache.beam.module' }
+apply plugin: 'maven-publish'
```
Comment (johnjcasey) [inline sdks/java/core/build.gradle:20]: Do you need this here?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Compression.java [hunk 3]
```diff
@@ -240,8 +244,66 @@ public WritableByteChannel writeCompressed(WritableByteChannel channel) throws I
       return Channels.newChannel(
           new SnappyCompressorOutputStream(Channels.newOutputStream(channel), uncompressedSize));
     }
+  },
+
+  /** Four MC Compression Splittable compressed file format leveraging LZ4
+   * Uses LZ4 Fast Compression*/
+  MC4_FAST(".4mc", ".4mc") {
```
Comment (johnjcasey) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Compression.java:251]: I believe we will have problems with detection with all of these having the same suffix configurations

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Compression.java [hunk 4]
```diff
@@ -240,8 +244,66 @@ public WritableByteChannel writeCompressed(WritableByteChannel channel) throws I
...
+    }
+
+    @Override
+    public ReadableByteChannel readDecompressed(ReadableByteChannel channel) {
+      return channel;
+    }
+  },
+  /** Four MC Compression Splittable compressed file format leveraging LZ4
+   * Uses LZ4 MC Compression*/
+  MC4_MEDIUM(".4mc", ".4mc") {
```
Comment (johnjcasey) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Compression.java:265]: Please add tests for these compression formats. See CompressedSourceTest for similar tests.

**Review by johnjcasey — CHANGES_REQUESTED** [review]

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Compression.java (hunk 3 above)
Comment (johnjcasey) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/Compression.java:252]: can you show your experiments verifying that many types with the same exception doesn't break?

**Review by johnjcasey — CHANGES_REQUESTED** [review]

**Comment by sundarrajanraman** [comment]
Added the Junits that would test that .4mc files reading would not break the auto detection.

**Comment by johnjcasey** [comment]
Run Website_Stage_GCS PreCommit

**Comment by johnjcasey** [comment]
Run Website PreCommit

**Comment by johnjcasey** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by johnjcasey** [comment]
Run Java_Examples_Dataflow_Java17 PreCommit

**Comment by sundarrajanraman** [comment]
Run Java PreCommit

**Comment by sundarrajanraman** [comment]
Run Java PreCommit

**Comment by sundarrajanraman** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by sundarrajanraman** [comment]
Run Java PreCommit

**Comment by sundarrajanraman** [comment]
Run Java PreCommit

**Comment by sundarrajanraman** [comment]
Run Python_Integration PreCommit

**Comment by sundarrajanraman** [comment]
Run Python_Integration PreCommit

**Comment by sundarrajanraman** [comment]
Run Python_Integration PreCommit

**Comment by sundarrajanraman** [comment]
run java precommit

**Comment by sundarrajanraman** [comment]
Run Java_hadoop_IO_Direct PreCommit

**Comment by damccorm** [comment]
@johnjcasey would you mind taking this review forward?

**Comment by kennknowles** [comment]
run java precommit

**Pull request closed without merging**